### PR TITLE
[OSD-19304] - service_log handler for sending requests for a specific cluster to the OCM API

### PR DESCRIPTION
**Pull request by xiaoyu74**
### What type of PR is this?
_(feature/refactor)_

**1. Create a Service Log Handler Using the OCM SDK:**
The `ServiceLogHandler` in `service_logs.go` uses the OCM SDK to manage service logs for clusters. It formats and sends log entries via the OCM API using the SDK's `LogEntry` and `ClusterLogsClient`.
```
request := h.ocm.ServiceLogs().V1().ClusterLogs().Add().Body(logEntry)
```

**2. Refactor the existing SendServiceLog function using the new created service log handler:** 
The existing `SendServiceLog()` function in `helper.go`, which originally interacted directly with the OCM API, is being refactored. Instead of making direct API calls, the function now uses the newly created `ServiceLogsHandler` implementing by the existing service log OCM SDK

**P.S.** 
In the initial SendServiceLog function, the `DocReferences` are included in the `sl` object, which is then converted to JSON and sent in the POST request to the OCM API. 
```
sl := ocm.ServiceLog{
		ServiceName:   consts.ServiceLogServiceName,
		ClusterUUID:   clusterID,
		Summary:       summary,
		InternalOnly:  false,
		Severity:      severity,
		LogType:       logType,
		DocReferences: references,
}
```

However, during my refactoring using existing OCM SDK types,  I found the current OCM SDK doesn't support for adding `references` to `LogEntry` (Also read the [OCM SDK doc](https://pkg.go.dev/github.com/openshift-online/ocm-sdk-go@v0.1.331/servicelogs/v1) but no functions/types for adding `references`).

So, I just used an alternative way in the refactored `SendServiceLogs()` function in `helper.go` to convert references to JSON and include them in the description.

### Which Jira/Github issue(s) this PR fixes?
https://issues.redhat.com/browse/OSD-19304

### CODE REVIEW

File: pkg/handlers/helper.go
Comment (hectorakemp) [inline pkg/handlers/helper.go:146]: Can we rename this variable to `clusterUUID`?

File: pkg/handlers/helper.go
Comment (hectorakemp) [inline pkg/handlers/helper.go:140]: Here too (`clusterID` becomes `clusterUUID`)

**Review by hectorakemp — CHANGES_REQUESTED** [review]
Very minor changes - but otherwise, looks awesome @xiaoyu74. 
- [ ] Could we add a test for the service log handler (copy format from https://github.com/openshift/ocm-agent/blob/e7734bbfb233b7c129063aae17919aeddb2155e5/pkg/handlers/upgrade_policies_test.go)
- [ ] Were you able to test it on a live stage cluster? Use `test/build-and-run.sh <cluster-id>` to make sure :)

File: pkg/handlers/helper.go
Comment (hectorakemp) [inline pkg/handlers/helper.go:142]: Also, a style suggestion - could we even move the whole method into `handlers/service_log`? Just to make it more organised.

Comment (xiaoyu74) [reply pkg/handlers/helper.go:142]: @hectorakemp - That's a good point. Actually, it's my intention to keep the `ServiceLogHandler` focusing solely on API communication. The `SendServiceLog` function, on the other hand, will handle the business logic, such as forming the appropriate summary and description based on the firing status.

As we expand the service log handler to include additional OCM API endpoints in the future, I expect that it continues to specialize in API interactions without delving into the business logic. WDYH?

Comment (hectorakemp) [reply pkg/handlers/helper.go:142]: Good point Shawn - I'm pretty new to OA - so I'm not sure what the overall architectural intentions of the operator are in the future. I personally don't mind either way. WDYT @bmeng ?

File: pkg/handlers/helper.go
Comment (xiaoyu74) [reply pkg/handlers/helper.go:140]: good catch, actually, I was just trying to build the same `LogEntry` pattern as in the original  `SendServiceLog()` function. 
```
sl := ocm.ServiceLog{
		ServiceName:   consts.ServiceLogServiceName,
		ClusterUUID:   clusterID,
		Summary:       summary,
		InternalOnly:  false,
		Severity:      severity,
		LogType:       logType,
		DocReferences: references,
}
```
But yeah, we should rename the variable to maintain consistency.

**Comment by xiaoyu74** [comment]
> Very minor changes - but otherwise, looks awesome @xiaoyu74.
> 
> * [ ]  Could we add a test for the service log handler (copy format from https://github.com/openshift/ocm-agent/blob/e7734bbfb233b7c129063aae17919aeddb2155e5/pkg/handlers/upgrade_policies_test.go)
> * [ ]  Were you able to test it on a live stage cluster? Use `test/build-and-run.sh <cluster-id>` to make sure :)

- Running the script against a staging cluster to build ocm-agent locally within my branch for service log handler

```
 ./build-and-run.sh 26livdquu06lqi60virhq2gskhonshkp

--- Fetching Cluster ID and creating temporary KUBECONFIG...

--- Building ocm-agent locally...
make: Entering directory '/home/sbai/osd_work/github_repo/ocm-agent'
?   	github.com/openshift/ocm-agent/cmd/ocm-agent	[no test files]
?   	github.com/openshift/ocm-agent/pkg/cli	[no test files]
?   	github.com/openshift/ocm-agent/pkg/cli/serve	[no test files]
?   	github.com/openshift/ocm-agent/pkg/config	[no test files]
?   	github.com/openshift/ocm-agent/pkg/consts	[no test files]
?   	github.com/openshift/ocm-agent/pkg/consts/test	[no test files]
ok  	github.com/openshift/ocm-agent/pkg/handlers	0.616s
?   	github.com/openshift/ocm-agent/pkg/handlers/mocks	[no test files]
?   	github.com/openshift/ocm-agent/pkg/k8s	[no test files]
?   	github.com/openshift/ocm-agent/pkg/logging	[no test files]
ok  	github.com/openshift/ocm-agent/pkg/metrics	(cached)
?   	github.com/openshift/ocm-agent/pkg/ocm	[no test files]
?   	github.com/openshift/ocm-agent/pkg/util/test/generated/mocks/client	[no test files]
?   	github.com/openshift/ocm-agent/pkg/util/test/mockgenerator/client	[no test files]
mkdir -p build/_output
GOOS=linux GOARCH=amd64 CGO_ENABLED=1 GOEXPERIMENT=boringcrypto GOFLAGS= go build -gcflags="all=-trimpath=/home/sbai/go" -asmflags="all=-trimpath=/home/sbai/go" -tags="fips_enabled" -o build/_output/ocm-agent ./cmd/ocm-agent
make: Leaving directory '/home/sbai/osd_work/github_repo/ocm-agent'

--- Running ocm-agent locally...
(Keep this terminal open to follow log of ocm-agent. Open new terminal and run test scripts.)
Link: http://localhost:8081

***** Starting with FIPS crypto enabled *****
INFO[2023-11-16T13:17:43+13:00] Starting ocm-agent server                    
```

-  In another terminal to run the test
```
./test-alerts.sh yury-addon-test

- Running Pre Test Check to recreate the default ManagedNotification for successful tests...
W1116 13:17:58.433296 2389569 loader.go:221] Config not found: /tmp/yury-addon-test-kubeconfig-temp
error: Missing or incomplete configuration info.  Please point to an existing, complete config file:


  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config

To view or setup config directly use the 'config' command.
W1116 13:17:58.466780 2389581 loader.go:221] Config not found: /tmp/yury-addon-test-kubeconfig-temp
error: Missing or incomplete configuration info.  Please point to an existing, complete config file:


  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config

To view or setup config directly use the 'config' command.

### TEST 1 - Send Service Log for a firing alert

- Command: curl -s -X POST http://localhost:8081/alertmanager-receiver -H 'Content-Type: application/json' -d @/tmp/firing-alert.json
- Output: {"Error":null,"Code":200,"Status":"ok"}
- Alert POSTed successfully to ocm-agent!

TEST FAILED = Expected SL count: 188, Got SL count: 187.

### TEST 2 - Do not send Service Log again for the same firing alert for same day

- Command: curl -s -X POST http://localhost:8081/alertmanager-receiver -H 'Content-Type: application/json' -d @/tmp/firing-alert.json
- Output: {"Error":null,"Code":200,"Status":"ok"}
- Alert POSTed successfully to ocm-agent!

TEST PASSED = Expected SL count: 187, Got SL count: 187.

### TEST 3 - Send Servicelog for resolved alert

- Command: curl -s -X POST http://localhost:8081/alertmanager-receiver -H 'Content-Type: application/json' -d @/tmp/resolved-alert.json
- Output: {"Error":null,"Code":200,"Status":"ok"}
- Alert POSTed successfully to ocm-agent!

TEST FAILED = Expected SL count: 188, Got SL count: 187.
```
Looks the count passed 187 out of 188, but it might not be the issue of my PR since I got the same results to build `ocm-agent` locally within the current master branch. cc: @hectorakemp @bmeng

**Review by hectorakemp — APPROVED** [review]
Great work @xiaoyu74 - I'm happy with this :+1:

File: pkg/handlers/helper.go
Comment (bmeng) [reply pkg/handlers/helper.go:142]: I have no preference here, but seems this kind of make the `helper` itself a bit ambiguous.
It looks to me move the service log related functions to the `service_logs.go` looks more clear

File: pkg/handlers/helper.go
Comment (Nikokolas3270) [inline pkg/handlers/helper.go:140]: What do you think of outsourcing the aggregation of those parameters to a `slv1.LogEntry` in a separate method and have the `SendServiceLog` method directly consume a `slv1.LogEntry`?

This is what I am doing in [my PR](https://github.com/openshift/ocm-agent/pull/68/files#diff-d8a640025cf97351b951865cfa2c487e92267cac2b507b583d6f35c056c1525fR28), except that I was still using type `ocm.ServiceLog` in place of `slv1.LogEntry`.

Indeed, with mocking in mind, `SendServiceLog` should do as less work as possible in order to get the unit tests cover as much code as possible.
Indeed in [my PR](https://github.com/openshift/ocm-agent/pull/68/files#diff-d8a640025cf97351b951865cfa2c487e92267cac2b507b583d6f35c056c1525fR55), I am adding the possibility to replace place holders with alert labels & annotations in the service log summary & description; this must be done **before** calling `SendServiceLog` as this method is designed to be mocked/replaced in unit tests.

Now of course, if `SendServiceLog` directly consumes a `slv1.LogEntry`, then it is probably just an empty wrapper around `serviceLogsHandler.PostServiceLog`. This may still have some value if we want to mock `SendServiceLog` but not mock `serviceLogsHandler.PostServiceLog`... however why not change the paradigm to say that it is `serviceLogsHandler.PostServiceLog` which should be mocked.

Comment (Nikokolas3270) [reply pkg/handlers/helper.go:140]: Note that I can take care of changing the method signature while dealing with my change.

File: pkg/handlers/helper.go
Comment (xiaoyu74) [reply pkg/handlers/helper.go:142]: > I have no preference here, but seems this kind of make the `helper` itself a bit ambiguous. It looks to me move the service log related functions to the `service_logs.go` looks more clear

I was thinking to keep a clear distinction between API communication and business logic to make `SendServiceLog()` separate from the `handlers/service_logs.go`. But your suggestion to have service log related functions in one place also makes sense. I will update accordingly then.

File: pkg/handlers/helper.go
Comment (xiaoyu74) [reply pkg/handlers/helper.go:140]: Thanks you @Nikokolas3270 - To have the construction of `slv1.LogEntry` separated from the SendServiceLog function would indeed simplify the `SendServiceLog` function, making it easier to mock and test. 

However, considering your PR is in progress that may handle these changes, may be better I keep the current implementation for now to avoids conflicting changes?

**Comment by bmeng** [comment]
@xiaoyu74 Could you please try to squash the commits?

File: pkg/handlers/helper.go
Comment (Nikokolas3270) [reply pkg/handlers/helper.go]: Hello, sure lets keep it that way, we can always come back on that later :)

**Review by hectorakemp — APPROVED** [review]

**Comment by bmeng** [comment]
/lgtm
/hold
leave to hulk to make the final call

**Comment by ravitri** [comment]
/unhold

Thanks a lot all !

**Pull request merged**
